### PR TITLE
Refactor main loop and fix looping bug in measurements.

### PR DESCRIPTION
**Pull request by grybmadsci**
Refactored the main loop to be a bit more readable, no logic changes.

Also fixed Measurement.Validate() being called multiple times when loop testing by making a deepcopy of the Measurement object when we create the measurements.Collection().

### CODE REVIEW

File: openhtf/exe/__init__.py [hunk 1]
```diff
@@ -154,17 +89,95 @@ def Wait(self):
...
+    if not self.test.loop:
+      self._RunTestLoop.once(self)
+    else:
+      self._RunTestLoop()
+
+  @threads.Loop
+  def _RunTestLoop(self):
+    """Handles one whole test from start to finish.
+
+    When this finishes, the parent loops back around and calls us again.
```
Comment (jettisonjoe) [inline openhtf/exe/__init__.py]: This sentence of the docstring is no longer accurate (nor was it in the previous iteration of the code for that matter). We should probably update or remove it.

File: openhtf/util/validators.py [hunk 2]
```diff
@@ -107,11 +113,15 @@ class _MatchesRegex(object):
...
   
     def __init__(self, re_module, regex):
+      self._re_module = re_module
       self._compiled = re_module.compile(regex)
       self._regex = regex
   
     def __call__(self, value):
       return self._compiled.match(str(value)) is not None
+
+    def __deepcopy__(self, memo):
```
Comment (jettisonjoe) [inline openhtf/util/validators.py]: ```memo``` is unused in the function and so should be named ```_``` or ```dummy_memo``` to pass lint.

Comment (grybmadsci) [inline openhtf/util/validators.py]: Done.

File: openhtf/exe/__init__.py (hunk 1 above)
Comment (grybmadsci) [inline openhtf/exe/__init__.py]: Done.  Updated to include the semantics of @threads.Loop to save someone the extra level of indirection in doc-hunting.

**Comment by jettisonjoe** [comment]
LGTM

**Comment by jettisonjoe** [comment]
Merging this now, but reserving the right to discuss the placement of outcomes at greater length later.

**Pull request merged**